### PR TITLE
refact: Decide API

### PR DESCRIPTION
**Pull request by oakbani**
Summary
-------
This PR refactors and fixes the decide implementation in another branch. 
**todo: reasons work. It will be done separately**. 

Test plan
---------
- Unit Tests added. 
- Manually verified with Decide API tests in FSC

Issues
------

### CODE REVIEW

**Review by jaeopt — CHANGES_REQUESTED** [review]
Changes suggested. 
I also see some comments for previous PRs (#305, #309) not addressed yet. Can you fix them as well?

File: optimizely/optimizely.py [hunk 1]
```diff
@@ -89,7 +87,13 @@ def __init__(
         self.event_processor = event_processor or ForwardingEventProcessor(
             self.event_dispatcher, logger=self.logger, notification_center=self.notification_center,
         )
-        self.default_decisions = default_decisions or []
+
+        if default_decide_options is None:
+            self.default_decide_options = []
+
```
[the remark this answers is not in the document]
Comment (oakbani) [reply optimizely/optimizely.py:93]: I have already placed another check below which ensures that it must be a list when non-null

File: optimizely/optimizely.py [hunk 2]
```diff
@@ -1120,7 +1100,7 @@ def decide_all(self, user_context, decide_options=None):
 
         # check if SDK is ready
         if not self.is_valid:
-            self.logger.error(enums.Errors.INVALID_PROJECT_CONFIG.format('decide_all'))
+            self.logger.error(enums.Errors.INVALID_OPTIMIZELY.format('decide_all'))
```
[the remark this answers is not in the document]
Comment (oakbani) [reply optimizely/optimizely.py:1103]: it already does. Looks like github is not showing you the updated diff.

**Comment by oakbani** [comment]
@jaeopt  I have tried to address all of your comments both in this PR and the rest. I am yet to take a final look on the unit tests. You may re-review all new changes besides the tests file.

**Review by jaeopt — CHANGES_REQUESTED** [review]
A couple of changes suggested. Other than that, all looks good.
I didn't review changes in unit tests. I'll take a look at them with your remaining fixes.

**Comment by oakbani** [comment]
@jaeopt  Reasons work has been done. Unit testing is in progress. You may review reasons related work for now.

**Review by jaeopt — CHANGES_REQUESTED** [review]
Overall it looks good! 
Some small changes suggested.

File: tests/test_user_context.py [hunk 3]
```diff
@@ -418,3 +419,697 @@ def test_decide_feature_rollout__send_flag_decision_false(self):
...
+            mock_variation,
+            expected.flag_key,
+            expected.rule_key,
+            'feature-test',
+            expected.enabled,
+            'test_user',
+            {'browser': 'chrome'}
+        )
+
+    def test_decide__option__include_reasons__feature_test(self):
```
Comment (jaeopt) [inline tests/test_user_context.py:750]: These tests look good.
Can we add a few more tests validating "reasons" messages for:
- hit on everyone-else rule
- hit on none of the rules
- hit variation from "get_forced_variation"
- hit from "white_list"
- hit from user-profile-service
- any other important cases?

**Review by jaeopt — CHANGES_REQUESTED** [review]
Looks great!
Suggested a few more tests for "reasons" validation

**Review by jaeopt — APPROVED** [review]
LGTM

**Pull request merged**